### ISSUE TITLE
Any Way To Avoid Telegram Flood Prevention ?

### ISSUE THREAD

**Issue by FarhadMohseni** [sochix/TLSharp#474]
I Sending Massive Messages To Telegram But I Get Flood Prevention And I Have to wait. 

I Used Thread.Sleep() Between My Messages But I Get Same Results . Any Way To Avoid This?

**Comment by mehdiyahyavi** [comment]
Me too, have this problem and thread.sleep can't help.
im using try-catch block and in catch block, using thread.sleep(exception.TimeToWait), but with this way, application freez for secound such as 200s.
any solution ... :-(

**Comment by mostafa8026** [comment]
Is wait for 9800 seconds reasonable?

**Comment by re-al-7** [comment]
I have no problems with the next code:

~~~csharp
var store = new FileSessionStore();
var client = new TelegramClient(CSettings.intTelegramApiId, CSettings.strTelegramApiHash, store, "session");
var connected = await client.ConnectAsync();
var strImagesDir = "PATH_TO_IMAGE_FOLDER";
if (connected)
{
	var estaRegistrado = await client.IsPhoneRegisteredAsync(CSettings.strTelegramNumber);
	if (estaRegistrado)
	{
		if (!client.IsUserAuthorized())
		{
			string hash = await client.SendCodeRequestAsync(CSettings.strTelegramNumber);

			Console.WriteLine("Codigo recibido:");
			var strCodigo = Console.ReadLine();

			var user = await client.MakeAuthAsync(CSettings.strTelegramNumber, hash, strCodigo);
			Console.WriteLine("Bienvenido " + user.username);
		}

		var result = await client.GetContactsAsync();
		var destino = result.users.lists
			.OfType<TLUser>()
			.FirstOrDefault(x => x.phone == CSettings.strTelegramPhone);

		if (destino != null)
		{
			string[] files = Directory.GetFiles(strImagesDir);
			foreach (string file in files)
			{			
				if (file.ToUpper().EndsWith(".JPG") || file.ToUpper().EndsWith(".JPEG") ||
					file.ToUpper().EndsWith(".PNG"))
				{
					if (iAvance % 5 == 0)
						Thread.Sleep(20 * 1000);

					//Telegram
					FileInfo info = new FileInfo(file);
					var archivo = await client.UploadFile(info.Name + "." + info.Extension, new StreamReader(file));
					await client.SendUploadedPhoto(new TLInputPeerUser() { user_id = destino.id }, archivo, txtTelCaption.Text);
				}
				iAvance++;
			}
			Console.WriteLine("Se han enviado las imagenes a Telegram");
		}
	}
}   
~~~

**Comment by mostafa8026** [comment]
Flood prevention. Telegram now requires your program to do requests again only after 9792 seconds have passed (TimeToWait property). If you think the culprit of this problem may lie in TLSharp's implementation, open a Github issue please.
!!!!

**Comment by knocte** [comment]
@mostafa8026 do you think this problem lies in TLSharp? If yes, explain why, please.

**Comment by dozham** [comment]
Obviously it's a limitation imposed by Telegram API, in order to prevent non-human users from causing chaos and anarchy in the network.
Although the details about this limitation hasn't been officially specified, it seems the only solution is to find a way (maybe waiting longer between requests or distributing them etc.) to send less requests in a unit of time overall.

**Comment by hosseinGanjyar** [comment]
every body....solved this problem? I've got same issue... I call every 20 seconds a method, after 30 min show flood prevention. :(

**Comment by knocte** [comment]
@hosseinGanjyar you know, the telegram official Windows client is opensource too, and that one doesn't suffer these problems I guess. You could debug it, and try to figure out what's the difference between its implementation and TLSharp? (Or your app which consumes TLSharp API)

**Comment by hosseinGanjyar** [comment]
https://github.com/sochix/TLSharp/issues/525

**Comment by hosseinGanjyar** [comment]
@knocte I don't understand. Please tell me telegram Id for contact to U.
thnks

**Comment by mHCio** [comment]
Hi, any solution found here ?
Thanks

**Comment by mHCio** [comment]
@FarhadMohseni  did you found a solution for your problem ?

**Comment by FarhadMohseni** [comment]
No,This is a Telegram Api Limitation.

**Comment by hosseinGanjyar** [comment]
Every body....I solved this problem...


> **thread.sleep for random time between 60-120 seconds for any call service.**


I call a service over 1000 times with this method.

**Comment by HarmenJeurink** [comment]
I already have this while debugging login / auth issues. I have to wait 77943 seconds, that is over 21 hours. I have no loops, just manual starts, less than 100.

**Comment by ahasna** [comment]
@hosseinGanjyar 
It is great that you have found a solution to this ... if I am understanding correctly there should be a random waiting time between requests? Right?
Because I am using a fixed waiting time but still having the issue.
Thank you

**Comment by hosseinGanjyar** [comment]
@ahasna I work correctly with **random waiting time between requests**.

**Comment by ahasna** [comment]
@hosseinGanjyar Thank you! 
is there a minimum amount for the random waiting?

**Comment by hosseinGanjyar** [comment]
@ahasna 60-120 sec

**Comment by ReaGet** [comment]
@hosseinGanjyar Hello! I made this delay, but I still have this error. Why?

**Comment by hosseinGanjyar** [comment]
@ReaGet send me code
hossein.ganjyar@gmail.com

**Comment by hosseinGanjyar** [comment]
Hello every ones...
I received many mails & messages from people every days for tell a way for ignore **flood** warning for access data on **Telegram**. So, I've solution (no hack, no bad way..) , **just solution**, for access to telegram data **without flood**,bc their app is open source. 
My solution is 100% worked and I spend a lot of times for discovery that, So I can't share it for public because I work hardly. 
If everybody want to have it, so should give me wages.
My telegram id is: **@mh_gj**, Contact to me.
Thanks.

**Comment by bieladeka** [comment]
I have to wait about 10 hours.... Did anyone find a solution?

**Comment by digidecor3d** [comment]
> CSettings
 what is the CSettings ?

**Comment by ahmadfarrokh** [comment]
https://stackoverflow.com/questions/47389575/telegram-api-development-tools-limits?rq=1
answer of this link solve this issue

**Comment by eordukaya** [comment]
> 
> 
> Thanks....I solved this problem...

How did you solve. can you share. we have this error.

**Comment by ahmadfarrokh** [comment]
var store = new CustomSessionStore();
            client = new TelegramClient(apiId, "apiHash", store as
ISessionStore, "session", new
TLSharp.Core.Network.TcpClientConnectionHandler(conectarProxy));
            await client.ConnectAsync();
            if (client.IsConnected)
            {

                if (await client.IsPhoneRegisteredAsync(countryPrefix +
phoneNumber))
                {
                    if (!client.IsUserAuthorized())
                    {
                        hash = await
client.SendCodeRequestAsync(countryPrefix + phoneNumber); //request for
session creating
                        var user = await client.MakeAuthAsync(countryPrefix
+ phoneNumber, hash, "xxxxx");
                    }
                }

                resultContact = await client.GetContactsAsync();

            }
////////////////////////////////
public class CustomSessionStore : ISessionStore
    {
        public void Save(Session session)
        {
            var dir = Path.Combine(AppDomain.CurrentDomain.BaseDirectory,
"Sessions");

            if (!Directory.Exists(dir))
            {
                Directory.CreateDirectory(dir);
            }

            var file = Path.Combine(dir, "{0}.dat");

            using (FileStream fileStream = new
FileStream(string.Format(file, (object)session.SessionUserId),
FileMode.OpenOrCreate))
            {
                byte[] bytes = session.ToBytes();
                fileStream.Write(bytes, 0, bytes.Length);
            }
        }
///////////////////////////////////////////////////////////////////////////////////////////////
         private TcpClient conectarProxy(string httpProxyHost =
"127.0.0.1", int httpProxyPort = 8580)
        {
            var url = "http://" + httpProxyHost + ":" + httpProxyPort;
            var proxyUrl = WebRequest.DefaultWebProxy.GetProxy(new
Uri(url));
            WebResponse response = null;
            var tentativas = 10;

            while (tentativas >= 0)
            {
                var request = (HttpWebRequest)WebRequest.Create(url);
                request.KeepAlive = true;
                var webProxy = new WebProxy(proxyUrl);
                request.Proxy = webProxy;
                request.Method = "CONNECT";
                request.Timeout = 3000;

                tentativas--;
                try
                {
                    response = request.GetResponse();
                    break;
                }
                catch (Exception ex)
                {
                    if (tentativas >= 0 && ex.Message.Equals("The operation
has timed out", StringComparison.InvariantCultureIgnoreCase))
                    {
                        Console.WriteLine("Ocorreu timeout ao tentar se
conectar pelo proxy.");
                    }
                    else
                    {
                        throw new Exception("Algo deu errado", ex);
                    }
                }
            }
            var responseStream = response.GetResponseStream();
            Debug.Assert(responseStream != null);

            const BindingFlags Flags = BindingFlags.NonPublic |
BindingFlags.Instance;

            var rsType = responseStream.GetType();
            var connectionProperty = rsType.GetProperty("Connection",
Flags);

            var connection = connectionProperty.GetValue(responseStream,
null);
            var connectionType = connection.GetType();
            var networkStreamProperty =
connectionType.GetProperty("NetworkStream", Flags);

            var networkStream = networkStreamProperty.GetValue(connection,
null);
            var nsType = networkStream.GetType();
            var socketProperty = nsType.GetProperty("Socket", Flags);
            var socket = (Socket)socketProperty.GetValue(networkStream,
null);

            return new TcpClient { Client = socket };
        }

**Comment by hosseinGanjyar** [comment]
> > Thanks....I solved this problem...
> 
> How did you solve. can you share. we have this error.

Please wait for create new issue by me soon...

**Comment by knocte** [comment]
Why a new issue?

**Comment by hosseinGanjyar** [comment]
Hi every body...You can ask your questions in this group:
https://t.me/joinchat/JRwh2w0A-7YlrMT_V4KhFQ

**Comment by apexkid** [comment]
One way to circumvent this problem is to create multiple telegram clients from different phone numbers. Then call the API in round robin fashion from each client so that requests are distributed evenly

**Comment by hosseinGanjyar** [comment]
I have got an idea and solution. How many timesdo you call an API?
And which one of API? (name)
Please share sample code. I will help you.

**Comment by ajit116** [comment]
> Me too, have this problem and thread.sleep can't help.
> I'm using try-catch block and in catch block, using thread.sleep(exception.TimeToWait), but with this way, application freez for second such as 200s.
> No solution ... :-(
> 
> can you tel me, when u get this error ? in sendmessage method I'm try to send to 100 users successful without error.

I got 'Getting Flood Error from telegram. Script is stopping now. Please try again after some time.' after only 4 message and then after waiting for hours still not able to send. Any suggestion

**Comment by arpanjossan46** [comment]
Can u tell me how to use this . Bcoz i use another  script i am getting flood error

**Comment by thepasterover** [comment]
Has anyone found a optimisation time to avoid the PEER FLOOD ERROR?

**Comment by mostafa8026** [comment]
There is not such thing. It is completely on telegram algorithm that manage this time. you have to write the program for a general use, not spam ;)

**Comment by Mongo71** [comment]
You will not get flood errors with these software's..

Buy Our Telegram Groups/Channels Scraper & Adder Software..
Buy Our Telegram Groups/Channels Bulk Private Direct Members Messenger/Sender Software.
Watch Our Videos..
Telegram Groups/Channels Scraper & Adder Software..
https://www.youtube.com/watch?v=pL8dO_1QVf0

Telegram Groups/Channels Bulk Private Direct Members Messenger/Sender Software.
https://www.youtube.com/watch?v=t3j3N4xYwpA

Order From Our Website..
http://telegram-groups-scraper-adder.com/

**Comment by Mongo71** [comment]
OUR SCRIPTS DO NOT HAVE 'PEER FLOOD ERRORS' LIKE ALL THE OTHER SCRIPTS YOU CAN GET ONLINE. 
WE FIX THAT ON OUR SCRIPTS..
http://telegram-groups-scraper-adder.com/

**Comment by Andrey-Smolensky** [comment]
Как я понял, дело не в api Телеги, и не бан по ip, я менял номер телефона, и ip это не помогает, и остается одна версия, это сама библиотека делает ограничения, путем записи видимо в реестр, либо еще как то.

**Comment by D4v1dH03** [comment]
@sochix Recommand locking this issue to prevent further spam.
